### PR TITLE
Wildcards

### PR DESCRIPTION
**Pull request by postsolar**
Need #21 merged first. This branch is based on the latest commit on that branch.
This branch starts at [a04b227](https://github.com/Maskhjarna/tree-sitter-purescript/pull/25/commits/a04b2276c8b75336507a3e903ffcb4aaf902c846).
See tests for changes.
I also removed quite a bit of Haskell-related code.
I also might have added some formatting in `id.js` and `utils.js`, if it's a problem I can remove it.

This concerns a lot of modules so just unit tests are not enough. I'd welcome running a build on real code and reporting what I broke.

### CODE REVIEW

**Comment by Maskhjarna** [comment]
Something is off about parentheses around expressions.
```purescript
module Test where

f :: Array Int -> Array Int -> Array Int
f a b c = (a <> b <> c)
```
gives
```
(purescript [0, 0] - [4, 0]
  name: (qualified_module [0, 7] - [0, 11]
    (module [0, 7] - [0, 11]))
  (where [0, 12] - [0, 17])
  (signature [2, 0] - [2, 40]
    name: (variable [2, 0] - [2, 1])
    type: (fun [2, 5] - [2, 40]
      (type_apply [2, 5] - [2, 14]
        (type_name [2, 5] - [2, 10]
          (type [2, 5] - [2, 10]))
        (type_name [2, 11] - [2, 14]
          (type [2, 11] - [2, 14])))
      (fun [2, 18] - [2, 40]
        (type_apply [2, 18] - [2, 27]
          (type_name [2, 18] - [2, 23]
            (type [2, 18] - [2, 23]))
          (type_name [2, 24] - [2, 27]
            (type [2, 24] - [2, 27])))
        (type_apply [2, 31] - [2, 40]
          (type_name [2, 31] - [2, 36]
            (type [2, 31] - [2, 36]))
          (type_name [2, 37] - [2, 40]
            (type [2, 37] - [2, 40]))))))
  (function [3, 0] - [3, 23]
    name: (variable [3, 0] - [3, 1])
    patterns: (patterns [3, 2] - [3, 7]
      (pat_name [3, 2] - [3, 3]
        (variable [3, 2] - [3, 3]))
      (pat_name [3, 4] - [3, 5]
        (variable [3, 4] - [3, 5]))
      (pat_name [3, 6] - [3, 7]
        (variable [3, 6] - [3, 7])))
    rhs: (exp_parens [3, 10] - [3, 23]
      (exp_name [3, 11] - [3, 12]
        (variable [3, 11] - [3, 12]))
      (ERROR [3, 13] - [3, 22]
        (operator [3, 13] - [3, 15])
        (ERROR [3, 18] - [3, 20])))))
```

**Comment by postsolar** [comment]
> Something is off about parentheses around expressions.

Yes I noticed it yesterday too. I've been fiddling with precedence rules but I can't even figure out the source of the problem, so currently I'm just trying to rewrite `exp.js` more or less from scratch.

**Comment by postsolar** [comment]
It only affects operators in parentheses, while `(f a b)` and ``(a `f` b)`` and `(a + _)` and `(a `f` _)` work correctly.

File: grammar/exp.js [hunk 1]
```diff
@@ -282,11 +227,33 @@ module.exports = {
...
+   *
+   * Also see `exp_section_left` comment.
    */
-  exp_infix: $ => seq($._exp_infix, $._qop, $._lexp),
+  exp_infix: $ =>
+    prec(1, seq(
+      $._exp_infix,
+      $._q_op,
+      $._lexp
+    )),
```
Comment (Maskhjarna) [inline grammar/exp.js:238]:
```suggestion
  exp_infix: $ =>
    prec(1, seq(
      $._lexp,
      $._q_op,
      $._exp_infix
    )),
```
Letting this be right-associative seems to fix my issues. I don't know if it breaks anything else.

**Comment by Maskhjarna** [comment]
> It only affects operators in parentheses, while `(f a b)` and `` (a `f` b) `` and `(a + _)` and `(a `f` _)` work correctly.

``(a `f` b)`` seems to be broken aswell no? Wouldn't mirroring `exp_infix` be appropriate here, like so:
```
  exp_infix_ticked: $ =>
    prec.left(2, seq(
      $._lexp,
      ticked(choice($._qvarid, $._exp)),
      $._exp_infix
    )),
```

File: grammar/exp.js (hunk 1 above)
Comment (postsolar) [reply grammar/exp.js:238]: Hmm I'm getting no changes after a fresh clone of the branch at cc14b26. The tests still pass though.

**Comment by postsolar** [comment]
> `` (a `f` b) `` seems to be broken aswell no?

It works for me.

> Wouldn't mirroring exp_infix be appropriate here, like so

Nope. I tried this with and without your previous suggestion but some tests break and `(a + b)` still doesn't parse.

**Comment by postsolar** [comment]
Re: your suggestions. I now noticed that if I clone the grammar repo somewhere else so that there are two folders with the same grammar on my system, TS will always use the first one, even if I delete all cache and provide `--libdir ./` as an argument. So I haven't actually checked your suggestions. I will look into this later.

**Pull request closed without merging**